### PR TITLE
Make promote_op rely on Core.Inference.return_type

### LINKED ISSUE #17314: promote_op/broadcast should not call the function for one(T)
As was commented at https://github.com/JuliaLang/julia/pull/17300#issuecomment-231027267 and https://github.com/JuliaLang/julia/pull/17172#issuecomment-230975340, the `promote_op` function has a [fallback  method](https://github.com/JuliaLang/julia/blob/6ab1b197f42807362c77893ea112fbaaad510785/base/number.jl#L68) that calls `typeof(op(one(T)))`.  This is not really acceptable.

For one thing, `op(1)` may not even be valid—it may throw a `DomainError` e.g. if `op(x)=log(log(log(x)))`.   Even if `1` is a valid argument, it is unexpected and undesirable that `op` should be called without the user requesting it.  (e.g. imagine if `op` has side effects, or is simply an extremely expensive function.)

### PR DESCRIPTION
**Pull request by p-zubieta**
This should fix #17314 and the problem found in #17394

EDIT: This also removes all specialized `promote_op` methods, since the new ones should handle all previous cases.

CC @JeffBezanson, @nalimilan, @tkelman

### CODE REVIEW

**Comment by pabloferz** [comment]
@martinholters After this, you should be able to use `broadcast` in #17313 as you originally intended.

File: base/arraymath.jl [hunk 1]
```diff
@@ -84,14 +70,14 @@ end
...
-            F = similar(B, promote_array_type($f,typeof(A),typeof(B)))
+            F = similar(B, promote_op($f, typeof(A), T))
             for (iF, iB) in zip(eachindex(F), eachindex(B))
                 @inbounds F[iF] = ($f)(A, B[iB])
             end
             return F
         end
         function ($f){T}(A::AbstractArray{T}, B::Number)
-            F = similar(A, promote_array_type($f,typeof(A),typeof(B)))
+            F = similar(A, promote_op($f, T, typeof(B)))
```
Comment (martinholters) [inline base/arraymath.jl]: `promote_array_type` is responsible for e.g. `typeof(Float32[1,2,3]*1.0) == Vector{Float32}`, even though `Base.promote_op(*, Float32, typeof(1.0)) == Float64` and `typeof(Float32(1)*1.0) == Float64`. If you look at above machinery closer, you'll note that the result of `promote_op` is only used in a few cases, and that's for a reason.

**Comment by martinholters** [comment]
> After this, you should be able to use broadcast in #17313 as you originally intended.

How is that? The reason for not not using `broadcast` all the time in #17313 was that it incurs some constant overhead even if `promote_op` returns a perfectly valid result, ruining performance for small arrays. Actually, if this PR "fixes" `promote_op` to more often do the right thing (and makes it into 0.5), #17313 would be obsolete.

**Comment by pabloferz** [comment]
>  The reason for not not using broadcast all the time in #17313 was that it incurs some constant overhead even if promote_op returns a perfectly valid result, ruining performance for small arrays.

Sorry, I missed that in #17313.

> Actually, if this PR "fixes" promote_op to more often do the right thing (and makes it into 0.5), #17313 would be obsolete.

Well, this should make `promote_op` do the right thing more often.

**Comment by tkelman** [comment]
There's been resistance in the past to making program behavior (as opposed to just performance) depend on the results of inference.

**Comment by TotalVerb** [comment]
I understand that sentiment, but something needs to be done about `promote_op`. What about the compromise of using it only in the empty case?

**Comment by JeffBezanson** [comment]
The ideal situation is to use the inference result when it is concrete, and otherwise use the types of the actual values. Then there is only a problem in the empty case. Using the inferred type there is deemed acceptable for now.

**Comment by pabloferz** [comment]
@JeffBezanson Does the current proposal address your concerns (at least provisionally and until we have a better mechanism)?

Tests are now passing on Travis (at the moment AV has a very long queue) and I don't think this should affect performance in general. Of course, that should be checked.

File: test/broadcast.jl [hunk 2]
```diff
@@ -164,8 +164,6 @@ m = [1:2;]'
 @test @inferred([0,1.2].+reshape([0,-2],1,1,2)) == reshape([0 -2; 1.2 -0.8],2,1,2)
 rt = Base.return_types(.+, Tuple{Array{Float64, 3}, Array{Int, 1}})
 @test length(rt) == 1 && rt[1] == Array{Float64, 3}
-rt = Base.return_types(broadcast, Tuple{typeof(.+), Array{Float64, 3}, Array{Int, 3}})
-@test length(rt) == 1 && rt[1] == Array{Float64, 3}
```
Comment (tkelman) [inline test/broadcast.jl]: why is this deleted?

Comment (pabloferz) [inline test/broadcast.jl]: With the changes broadcast wouldn't be inferable (same as map) but it will return the correct type nevertheless. I guess I should change it to at least verify that.

**Comment by pabloferz** [comment]
The Travis failure seems to be due to #16537 (see #17409)

**Comment by tkelman** [comment]
> otherwise use the types of the actual values

This isn't doing the latter, it's returning `Any` when inference doesn't return a leaf type?

**Comment by pabloferz** [comment]
@tkelman You're right, but I believe the cases where the type returned by `return_type` is not concrete won't be the ones where we care about performance, so we can use the fall back `broadcast` method.

Actually, this is more in analogy with the `_default_eltype` used by `map` (through, the `collect` methods). So in a way `broadcast` and `map` would behave alike with this changes (this was what @stevengj was advocating IIUC).

If the first statement above is not accurate I can change the PR to accommodate Jeff's initial comment.

**Comment by yuyichao** [comment]
I'll be funny if we introduce another type inference dependent behavior for non-empty cases right after we worked really hard to fix one....

**Comment by stevengj** [comment]
No, what I was advocating was what @JeffBezanson said.   First you try inference.  If that doesn't return a concrete type, and the output is non-empty, you use the union of the types of the actual values to get the type.  Only if the output is empty would you ever use a non-concrete inferred type.

**Comment by stevengj** [comment]
My understanding is that @JeffBezanson already implemented getting the type from the actual values in the non-empty case in #17172, so all we really need to do is to fix the fallback of `promote_op`, no?

File: test/broadcast.jl [hunk 3]
```diff
@@ -223,12 +223,11 @@ let x = sin.(1:10)
     @test sin.(atan2.(x, 3.7)) == broadcast(x -> sin(atan2(x,3.7)), x)
     @test atan2.(x, 3.7) == broadcast(x -> atan2(x,3.7), x) == broadcast(atan2, x, 3.7)
 end
-# Use side effects to check for loop fusion.  Note that, due to #17314,
-# a broadcasted function is currently called an extra time with an argument 1.
+# Use side effects to check for loop fusion.
 let g = Int[]
     f17300(x) = begin; push!(g, x); x+1; end
```
Comment (stevengj) [inline test/broadcast.jl]: Probably this should be changed to e.g. `x+2`, since with `x+1` it will now give the same results whether the loops are fused or not.

File: test/broadcast.jl [hunk 4]
```diff
@@ -256,3 +255,10 @@ end
 let foo = [[1,2,3],[4,5,6],[7,8,9]]
     @test max.(foo...) == broadcast(max, foo...) == [7,8,9]
 end
+
+# Issue 17314
```
Comment (stevengj) [inline test/broadcast.jl]: Would be good to test a case where type-inference fails, e.g. `broadcast(x -> x < 0 ? false : x, 1:3)` should return `Int[1,2,3]`.

Comment (stevengj) [inline test/broadcast.jl]: And `broadcast(x -> x < 0 ? false : x, Int[])` should probably return the inferred type `Union{Bool,Int}[]`.

**Comment by pabloferz** [comment]
Ok. Seems there was a misunderstanding. This uses inference only when the returned type is concrete. When that is not the case `promote_eltype_op` will return `Any` and then `broadcast` will dispatch on a *method that uses the actual values* (that is what Jeff implemented in #17172) so the resulting `Array` will be of the same type that `map` would give.

I changed `promote_eltype_op` instead of `promote_op` because the latter is used by many other methods. For these methods, the current behaviour of `promote_op` is so far the best option until we find a better promotion mechanism for them.

@stevengj I added the tests you suggested.

**Comment by pabloferz** [comment]
Of course, I can change `promote_op` instead of `promote_eltype_op` if that is preferred. An example of a method that would return different results with the current proposal is the following

```Complex[1 1; 1 1] * Complex{Float64}[1 1; 1 1]```

which gives an `Array{Complex,2}`, but with the change would give an `Array{Any,2}`

**Comment by martinholters** [comment]
In the long run, I think, we should not try "fixing" `promote_op` and friends - it's just a broken concept. Instead, we should aim at removing every use of them, instead working out the result type using the actual input values. Inference might be used to fast-track this by fixing the result type up front, and for fixing the type of the empty input case.

**Comment by pabloferz** [comment]
@martinholters I agree that `promote_op` should eventually be improved or removed in methods other than `broadcast`. That is why I only modified `promote_eltype_op`.

Until the type system gets redesigned and we have traits, this seems to me the best option to get `broadcast` methods working properly without disrupting the behaviour or linear algebra methods and arithmetic operations on `Array`s.

**Comment by martinholters** [comment]
> Until the type system gets redesigned and we have traits, this seems to me the best option

Huh, how do traits help here? Please do not suggest reworking the current `promote_op` mess into some future trait syntax - better syntax will not help with the scalability and composability issues.

**Comment by pabloferz** [comment]
> Huh, how do traits help here? Please do not suggest reworking the current promote_op mess into some future trait syntax - better syntax will not help with the scalability and composability issues.

I'm not suggesting that. I was thinking in how traits could help to handle choosing a method where inference will help you to fix the result up-front and vs. when it won't.

**Comment by pabloferz** [comment]
All right. The new solution gets rid off all the specialized `promote_op` methods. With these changes `promote_op` shouldn't be the same broken-concept as before and should help in the future to identify some type instabilities for certain numeric methods (as already happened for the current PR).

I believe this should be ready once the tests pass. Also, I believe this is good candidate for merging during the 0.5-RC cycle or to be backported on 0.5.1.

**Comment by pabloferz** [comment]
I believe the Travis failure is unrelated.

File: base/complex.jl [hunk 5]
```diff
@@ -467,7 +467,7 @@ function ^{T<:AbstractFloat}(z::Complex{T}, p::Complex{T})
         elseif isnan(y) && isinf(x)
             y = copysign(zero(T), y)
         end
-        Complex(x,y)
+        Complex{T}(x, y)
```
Comment (nalimilan) [inline base/complex.jl]: Probably better use a return type declaration if you need to change these.

File: base/irrationals.jl [hunk 6]
```diff
@@ -94,6 +87,7 @@ hash(x::Irrational, h::UInt) = 3*object_id(x) - h
 for op in Symbol[:+, :-, :*, :/, :^]
     @eval $op(x::Irrational, y::Irrational) = $op(Float64(x),Float64(y))
 end
+*(x::Bool, y::Irrational) = ifelse(x, float(y), 0.0)
```
Comment (nalimilan) [inline base/irrationals.jl]: To be sure this is type-stable, why not `Float64(y)` instead of `float(y)`?

**Comment by tkelman** [comment]
this is still returning Any if inference doesn't return a leaf type, isn't it?

**Comment by martinholters** [comment]
The current proposal looks like running head on against @yuyichao's concern in https://github.com/JuliaLang/julia/pull/17389#issuecomment-233132876. Let me also repeat myself: We should not try "fixing" `promote_op`, we should strive to avoid using it. Might be a bit much to completely do so in one strike, but starting with `broadcast` seemed like a reasonable idea to see how it goes.

**Comment by tkelman** [comment]
Some of the type-stability fixes in here look like they could be independently applicable and quite useful, so might be good to separate them out from the inference-using parts if possible?

**Comment by pabloferz** [comment]
> The current proposal looks like running head on against @yuyichao's concern in #17389 (comment).

On the contrary, I believe this is aligned with that concern and also with this comment of yours

> Inference might be used to fast-track this by fixing the result type up front, and for fixing the type of the empty input case.

Maybe I just need to clarify some things (which I haven't been able to do so far). @JeffBezanson implemented a mechanism to do what you describe above in https://github.com/JuliaLang/julia/pull/16622/commits/928bcde3b78951f4fd145423e547b2377d03cb21. There is a function `_default_type` over that commit that relies on Inference to achieve that for maps and comprehensions.

In my view, `promote_op` should be to `broadcast` what `_default_type` is to `map`, which is what I actually implemented here (it might not seems obvious, but this is doing exactly the same). Jeff also implemented a broadcast method that will return an `Array` with an appropriate type for functions that can't be inferred in  #17172. For the later broadcast method you need `promote_op` to return `Any`, which again is what I have here. It is for these reasons that `promote_op` should only be used by broadcast (or methods that rely on it) the same way `_default_type` is only used for map and comprehensions.

Currently, `promote_op` is also used for some linear algebra methods that do either sum or multiplication of arrays, in which case `promote_op` is generally safe to use. In these cases it may be desirable in the future to remove the dependency on it but that should be done in another PR.

The only other place where `promote_op` was being used was for binary element-wise operations on numbers and arrays (the case of having arguments as arrays you already addressed it in #17313). This is also changed here to make sure we use `broadcast` when we can't infer the type. Eventually, we might seek to make all these to call broadcast directly, but we would have to take care of the overhead you found in #17313.

In summary, if you see `promote_op` for `broadcast` as the analogous of `_default_type` for `map`, then is my opinion that it should be fixed (that is what I did here). On the other hand, if look at it in other ways, then you might want to get rid off it.

**Comment by stevengj** [comment]
In other words, as I understand this patch, it only uses the type-inference result in `broadcast` if the type is concrete or if the output array is empty, similar to `map`.  This is safe and shouldn't (except in the empty output case, which has been discussed extensively elsewhere) introduce an inference-dependency into the output.

**Comment by pabloferz** [comment]
> In other words, as I understand this patch, it only uses the type-inference result in broadcast if the type is concrete or if the output array is empty, similar to map.

@stevengj Right, that's exactly the case.

**Comment by pabloferz** [comment]
I believe the Travis failure is #16091

**Comment by stevengj** [comment]
Restarting Travis.

**Comment by tkelman** [comment]
would returning Any have an effect on any of the packages that are currently calling this?

**Comment by pabloferz** [comment]
@tkelman If they defined custom `promote_op` methods, then, no (but eventually they should remove dependencies on it). If the were using it for functions that are type-stable they also would be safe. The main problem would be if there are uses of `promote_op` for operations on custom subtypes of `Number` that are not type stable.

But, I don't know how what packages rely on the method and how their are using it.

**Comment by tkelman** [comment]
https://gist.github.com/anonymous/c7bf0061887128f7941698a6807a5b2a

**Comment by stevengj** [comment]
Grr, `LoadError: error compiling mod2pi: could not load library "libopenspecfun"` must be unrelated.

**Comment by tkelman** [comment]
we should open a separate issue on that, I believe the rpath changes are responsible for that. Note that when you restart travis is overwrites the old log, so best to back it up to a gist first.

**Comment by pabloferz** [comment]
https://gist.github.com/anonymous/e2e4e558537b4836ffdbe47a0bc04454

**Comment by pabloferz** [comment]
Its a shame that this seems to trigger #16091 more often (even though it also might help to identify the root cause of that issue)

File: base/arraymath.jl [hunk 7]
```diff
@@ -35,53 +35,71 @@ function !(A::AbstractArray{Bool})
...
-@pure promote_array_type{S, A, P}(F, ::Type{S}, ::Type{A}, ::Type{P}) = P
-@pure promote_array_type{S<:Real, A<:AbstractFloat, P}(F, ::Type{S}, ::Type{A}, ::Type{P}) = A
-@pure promote_array_type{S<:Integer, A<:Integer, P}(F::typeof(./), ::Type{S}, ::Type{A}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, A<:Integer, P}(F::typeof(.\), ::Type{S}, ::Type{A}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, A<:Integer, P}(F, ::Type{S}, ::Type{A}, ::Type{P}) = A
-@pure promote_array_type{S<:Integer, P}(F::typeof(./), ::Type{S}, ::Type{Bool}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, P}(F::typeof(.\), ::Type{S}, ::Type{Bool}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, P}(F, ::Type{S}, ::Type{Bool}, ::Type{P}) = P
+
+@pure _promote_type(cond, T) = cond ? T : (T <: Complex ? Complex{Int} : Int)
```
Comment (tkelman) [inline base/arraymath.jl]: why the Int fallbacks here?

File: base/promotion.jl [hunk 8]
```diff
@@ -217,13 +217,27 @@ max(x::Real, y::Real) = max(promote(x,y)...)
 min(x::Real, y::Real) = min(promote(x,y)...)
 minmax(x::Real, y::Real) = minmax(promote(x, y)...)
 
-# "Promotion" that takes a function into account. You can override this
-# as needed. For example, if you need to provide a custom result type
-# for the multiplication of two types,
-#   promote_op{R<:MyType,S<:MyType}(::typeof(*), ::Type{R}, ::Type{S}) = MyType{multype(R,S)}
+# "Promotion" that takes a function into account. These are meant to be used
+# mainly by broadcast methods, so it is adviced against overriding them
```
Comment (tkelman) [inline base/promotion.jl]: advised (you can [ci skip] this)

File: base/arraymath.jl [hunk 9]
```diff
@@ -35,53 +35,71 @@ function !(A::AbstractArray{Bool})
...
+end
+
+@pure promote_array_type(F, ::Type, ::Type, T::Type) = T
+@pure promote_array_type{S<:Real, A<:AbstractFloat}(F, ::Type{S}, ::Type{A}, ::Type) = A
+@pure promote_array_type{S<:Integer, A<:Integer}(F, ::Type{S}, ::Type{A}, ::Type) = A
+@pure promote_array_type{S<:Integer, A<:Integer}(::typeof(./), ::Type{S}, ::Type{A}, T::Type) = T
+@pure promote_array_type{S<:Integer, A<:Integer}(::typeof(.\), ::Type{S}, ::Type{A}, T::Type) = T
+@pure promote_array_type{S<:Integer}(::typeof(./), ::Type{S}, ::Type{Bool}, T::Type) = T
+@pure promote_array_type{S<:Integer}(::typeof(.\), ::Type{S}, ::Type{Bool}, T::Type) = T
+@pure promote_array_type{S<:Integer}(F, ::Type{S}, ::Type{Bool}, T::Type) = T
```
Comment (vtjnash) [inline base/arraymath.jl]: there's no value in marking these `pure`. it's obvious that they return a single constant

File: base/promotion.jl [hunk 10]
```diff
@@ -217,13 +217,27 @@ max(x::Real, y::Real) = max(promote(x,y)...)
...
 promote_op{T}(::Type{T}, ::Any) = (@_pure_meta; T)
+promote_op{T}(::Type{T}, ::Type) = (@_pure_meta; T) # To handle ambiguities
+if isdefined(Core, :Inference)
+    function promote_op(op, T::Type)
+        G = Tuple{Generator{Tuple{T},typeof(op)}}
+        R = Core.Inference.return_type(first, G)
+        return isleaftype(R) ? R : Any
+    end
+    function promote_op(op, R::Type, S::Type)
+        isleaftype(R) || isleaftype(S) || return typejoin(R, S)
```
Comment (martinholters) [inline base/promotion.jl]: This looks like a dangerous assumption. E.g. `promote_op(string, Number, Number)` should not return `Number`. Granted, `broadcast(string, Number[1], Number[1])` seems rather useless, but there may well be better examples triggering the same problem.

File: base/promotion.jl [hunk 11]
```diff
@@ -217,13 +217,27 @@ max(x::Real, y::Real) = max(promote(x,y)...)
...
+        F = typeof(a -> op(a...))
+        G = Tuple{Generator{Zip2{Tuple{R},Tuple{S}},F}}
+        T = Core.Inference.return_type(first, G)
```
Comment (martinholters) [inline base/promotion.jl]: Why wouldn't just `Core.Inference.return_type(op, (R,S))` work here?

**Comment by martinholters** [comment]
Wouldn't the simple
```Julia
function promote_op(op, argtypes...)
    T = Core.Inference.return_type(op, argtypes)
    return isleaftype(T) ? T : Any
end
```
and maybe
```Julia
function promote_op{T}(::Type{T}, ::Any...) = T
```
be enough for your purposes?

File: base/arraymath.jl (hunk 7 above)
Comment (pabloferz) [inline base/arraymath.jl]: This is for things like `1 * Number[1] == Number[1]` to work. Without this the `promote_op` below would return `Any` and then `broadcast` would return an `Array{Int}`.

File: base/promotion.jl (hunk 8 above)
Comment (pabloferz) [inline base/promotion.jl]: Do you mean, I should have a separate commit with this [ci skip]ed?

File: base/promotion.jl (hunk 10 above)
Comment (pabloferz) [inline base/promotion.jl]: I was hesitant about it. I wouldn't have it here except that there are some linear algebra methods that rely on it. I can take that out, but there might be people unsatisfied with some specific products of matrices returning `Array{Any}`.

File: base/promotion.jl (hunk 8 above)
Comment (tkelman) [inline base/promotion.jl]: If the CI queue is long and there are no more code changes to make then that can help, yeah (it can be squashed on merge). The queue is quiet right now though so doesn't make much difference.

**Comment by pabloferz** [comment]
@martinholters That was one of the first things I tried, just to find out that `promote_op` wouldn't be inferable that way.

File: base/promotion.jl (hunk 10 above)
Comment (martinholters) [inline base/promotion.jl]: There surely are. I guess this is where the "take the type of the actual elements" path would be needed. Then e.g. `Number[1] + Number[1]` would give `Int[2]` instead of `Number[2]` though, but that seems still better than `Any[2]`. (I guess this would be one of the critical situations?)

**Comment by martinholters** [comment]
But maybe still a good replacement for the `promote_op(::Any...) = (@_pure_meta; Any)` fallback?

File: base/promotion.jl (hunk 10 above)
Comment (nalimilan) [inline base/promotion.jl]: Returning `Int[2]` would be annoying if you later want to add non-integer values to that array. If the input eltype is `Number`, we have to respect this and not return a more precise type.

Comment (nalimilan) [inline base/promotion.jl]: Which methods rely on this? I thought the goal was to get rid of generic `promote_op` methods which do not take into account the actual return type of the method.

Comment (martinholters) [inline base/promotion.jl]: > If the input eltype is Number, we have to respect this and not return a more precise type.

While that sounds quite desirable, I don't see how that could be done without a mechanism like the current `promote_op` or relying on inference.

**Comment by stevengj** [comment]
@JeffBezanson, is this shaping up into a mergeable state?

File: base/promotion.jl (hunk 10 above)
Comment (stevengj) [inline base/promotion.jl]: @pabloferz, if specific linear-algebra methods rely on it, can't we add workarounds for those specific methods?

**Comment by andreasnoack** [comment]
@pabloferz Since it is now used in the linear algebra code, how should I think about `promote_binary_op`? I'm a bit confused by the name because `promote_op` used to for binary operators.

**Comment by pabloferz** [comment]
@andreasnoack `promote_binary_op` is not an appropriate name, it should be more something like `promote_numeric_op` (suggestions are welcomed). With these changes `promote_op` gives back the return type of the operation as long as it can be inferred or `Any` otherwise. For numeric methods on arrays we might want to preserve non-concrete containers types, e.g. `Number[1] * 1 == Number[1]`. The method here named `promote_binary_op` will do the same that `promote_op` plus trying to preserve non-concrete container types.

The ideal would be probably to also have linear algebra methods that figure the type from the values if the type of the resulting array cannot be inferred. If that is to be expected that can be probably done later. I believe that in general, the current proposal wouldn't affect the current behaviour of linear algebra methods that rely on `promote_op`.

**Comment by martinholters** [comment]
Considering the uses of `promote_op` in packages, would it make sense to have the `promote_op` that is to be used by `broadcast` to go by a different name and keep `promote_op` for what you have now called `promote_binary_op`?

**Comment by pabloferz** [comment]
@martinholters Yeah, that's probably a better idea.

File: base/arraymath.jl [hunk 12]
```diff
@@ -35,30 +35,31 @@ function !(A::AbstractArray{Bool})
...
-@pure promote_array_type{S<:Integer, A<:Integer, P}(F::typeof(.\), ::Type{S}, ::Type{A}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, A<:Integer, P}(F, ::Type{S}, ::Type{A}, ::Type{P}) = A
-@pure promote_array_type{S<:Integer, P}(F::typeof(./), ::Type{S}, ::Type{Bool}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, P}(F::typeof(.\), ::Type{S}, ::Type{Bool}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, P}(F, ::Type{S}, ::Type{Bool}, ::Type{P}) = P
 
-for f in (:+, :-, :div, :mod, :&, :|, :$)
-    @eval ($f){S,T}(A::AbstractArray{S}, B::AbstractArray{T}) =
-        _elementwise($f, A, B, promote_eltype_op($f, A, B))
+@pure _default_type(ilT, T) = ilT ? T : (T <: Complex ? Complex{Int} : Int)
```
Comment (vchuravy) [inline base/arraymath.jl]: Why does this default to `Int`? I would like to understand the behaviour and the result of `promote_binary_op(+,AbstractFloat, AbstractFloat) == Real` seems counter-intuitive.

Comment (pabloferz) [inline base/arraymath.jl]: This was in analogy with how `promote_op` used `one` to give a result for numeric operations. But, as you point out that is probably not what we want here. I'll try to fix it.

**Comment by stevengj** [comment]
If there is a `promote_op` that is just used for `broadcast`, it should presumably be `promote_broadcast_op`.

**Comment by pabloferz** [comment]
Well, `promote_eltype_op` is basically `promote_op` for broacast. Since, it is currently used in some packages I'll leave it like that (though it would be nice to rename it to `broadcast_type` analogous to `broadcast_shape` or `promote_broadcast_op` in the future).

**Comment by pabloferz** [comment]
Ok. I believe this addresses all the comments and concerns so far.

(The last build failure seems to be happening in other PRs too)

File: base/arraymath.jl [hunk 13]
```diff
@@ -35,30 +35,23 @@ function !(A::AbstractArray{Bool})
...
+    @eval ($f)(A::AbstractArray, B::AbstractArray) =
+        _elementwise($f, promote_op($f, eltype(A), eltype(B)), A, B)
 end
-function _elementwise{S,T}(op, A::AbstractArray{S}, B::AbstractArray{T}, ::Type{Any})
-    promote_shape(A,B) # check size compatibility
-    return broadcast(op, A, B)
-end
-function _elementwise{S,T,R}(op, A::AbstractArray{S}, B::AbstractArray{T}, ::Type{R})
-    F = similar(A, R, promote_shape(A,B))
+_elementwise(op, ::Type{Any}, A::AbstractArray, B::AbstractArray) = op.(A, B)
```
Comment (martinholters) [inline base/arraymath.jl]: That would allow e.g. `zeros(Foo,1,2) + zeros(Foo,2,1)` if `promote_op` yields `Any`, but give a `DimensionMismatch` error otherwise. Probably best to always error as we have now.

File: base/arraymath.jl [hunk 14]
```diff
@@ -67,15 +60,21 @@ end
 
 for f in (:.+, :.-, :.*, :./, :.\, :.^, :.÷, :.%, :.<<, :.>>, :div, :mod, :rem, :&, :|, :$)
     @eval begin
-        function ($f){T}(A::Number, B::AbstractArray{T})
-            F = similar(B, promote_array_type($f,typeof(A),typeof(B)))
+        function ($f)(A::Number, B::AbstractArray)
+            P = promote_op($f, typeof(A), eltype(B))
+            T = promote_array_type($f, typeof(A), eltype(B), P)
+            T === Any && return [($f)(A, b) for b in B]
+            F = similar(B, T)
```
Comment (martinholters) [inline base/arraymath.jl]: If `T==Any`, it will return an `Array`, otherwise it may not, depending on `similar`. But maybe we don't have to worry about that now. These changed will be obsoleted by the upcoming lowering of `a .op b` as `(op).(a,b)` anyway.

File: base/bitarray.jl [hunk 15]
```diff
@@ -1078,9 +1089,9 @@ function div(x::Bool, B::BitArray)
 end
 function div(x::Number, B::BitArray)
     all(B) || throw(DivideError())
-    pt = promote_array_type(div, typeof(x), BitArray)
     y = div(x, true)
-    reshape(pt[ y for i = 1:length(B) ], size(B))
+    T = typeof(y)
+    reshape(T[ y for i = 1:length(B) ], size(B))
```
Comment (martinholters) [inline base/bitarray.jl]: `fill(y, size(B))` or `[y for b in B]`?

File: base/arraymath.jl (hunk 14 above)
Comment (pabloferz) [inline base/arraymath.jl]: It will return an Array even if I'd use broadcast here (broadcast returns an Array too, but that is a separate issue).

Comment (martinholters) [inline base/arraymath.jl]: Yes, so once the dot-operators are turned into `broadcast`s we will _consistently_ have `Array` return types (unless we adapt `broadcast` as is already done to some extent for `SparseMatrixCSC`, which we probably should do). Having `promote_op` not only decide the element type, but also the container type _might_ be an issue, but only temporarily anyway.

Comment (pabloferz) [inline base/arraymath.jl]: Given that this issue didn't stop #17313 to get merged, I don't think it should hold down this one. Of course, as you point out, its temporal and should be fixed separately.

**Comment by pabloferz** [comment]
There has been a couple of questions over julia-users (https://groups.google.com/forum/#!topic/julia-users/nzKu4Adh0Tw and https://groups.google.com/forum/#!topic/julia-users/4o2wlCzTYhI) regarding some of the problems this PR fixes. Someone asked if this was going to make it into 0.5. Since, this is a bug fix I'd think it would pose a problem merging it. Of course, that is not up to me and I leave it consideration for the appropriate parties.

@tkelman I checked all packages that use these function and I submitted a couple of PRs that should prevent (or fix) the only two minor problems that would arise from merging this.

**Comment by tkelman** [comment]
Have the PR's been merged and tagged yet? What we could do is build test binaries that have this change included and run PackageEvaluator to see if there are any regressions, as I've been doing for backports.

**Comment by pabloferz** [comment]
https://github.com/JuliaStats/DataArrays.jl/issues/209 was merged but not tagged.
https://github.com/JuliaMatrices/BandedMatrices.jl/issues/10 is supposed to be merged after this is merged.

Anyway, I guess it's worth doing what you propose.

**Comment by stevengj** [comment]
@tkelman, also note that the problems being fixed by this PR are a regression from 0.4.

**Comment by tkelman** [comment]
I did a manual pkgeval run here overnight, will put up a comparison summary shortly.

**Comment by tkelman** [comment]
did I have a comment here several days ago that someone deleted?

**Comment by pabloferz** [comment]
I don't think so. What was the comment about?

**Comment by tkelman** [comment]
I could have sworn that a few days ago I said I didn't think we should merge this until after we branch, though that was before seeing the issues raised in the last few days, and I need to look at the pkgeval results.

**Comment by pabloferz** [comment]
This https://github.com/JuliaLang/julia/pull/16986#issuecomment-236014163? I guess you misplaced it.

**Comment by tkelman** [comment]
I was thinking of https://github.com/JuliaLang/julia/issues/16986, nevermind, sorry.

**Comment by tkelman** [comment]
Sorry for the delay, took me a while to get in front of a proper keyboard.

This turns out to be hugely breaking in packages: https://gist.github.com/8f19fefaf5021dbc076cc47b3ff8abd1

Could use a hand digging through those json logs (jq and/or JSON.jl are useful for this) to see if they're all the same failure though, maybe a re-run now that DataArrays has been tagged might fix them all?

**Comment by tkelman** [comment]
With DataArrays tagged it's a much smaller number of failures now. https://gist.github.com/5049d028f493ef2a9163a269f9a634fc
Just BandedMatrices (which you have a fix for), GaussianMixtures, and NearestNeighbors. Could you take a look at the other 2?

(Reactive has flaky tests, and apparently this PR actually fixes SampledSignals http://pkg.julialang.org/detail/SampledSignals.html)

**Comment by tkelman** [comment]
`signal (15): Terminated\nwhile loading no file, in expression starting on line 0\n\nsignal (15): Terminated\nwhile loading /home/vagrant/.julia/v0.5/NearestNeighbors/test/test_monkey.jl, in expression starting on line 6\nuv__epoll_wait`
I think that means NearestNeighbors' tests got so slower with this change and they timed out.

**Comment by KristofferC** [comment]
Maybe deprecation warnings? I fixed deprecation now so please try again.

**Comment by tkelman** [comment]
PkgEval runs on tagged package versions.

**Comment by KristofferC** [comment]
Ok, I tag then.

**Comment by pabloferz** [comment]
The NearestNeighbors one is fixed (locally) for me by fixing the deprecations as @KristofferC pointed out. 

On the other hand I'm not that sure how this is affecting GaussianMixtures, it passed the first time, but it did't passed the second (on a function that has nothing to do with the DataArrays). Also, it passes the tests on my machine. I suspect the memory used fluctuates around the edges of the memory of the testing machine, but I'm not sure.

**Comment by tkelman** [comment]
Ah right it's an `OutOfMemoryError()` so possibly unrelated. Do you see a noticeable difference in peak memory use on this branch vs master?

**Comment by pabloferz** [comment]
> Do you see a noticeable difference in peak memory use on this branch vs master?

I don't. But you already merged it, so I guess there is no problem.

**Comment by tkelman** [comment]
There might be a problem, but I'll take one fixed package (plus 2 about-to-be-fixed packages) in exchange for one broken one. cc @davidavdav

**Comment by timholy** [comment]
> Reactive has flaky tests

That might have been fixed by https://github.com/JuliaLang/Reactive.jl/commit/d7a2bd36ead56cf1e0bdad6b04af9e6a1803860b, which was tagged just a couple of days ago.

**Comment by tkelman** [comment]
Not really http://pkg.julialang.org/detail/Reactive.html

File: base/promotion.jl [hunk 16]
```diff
@@ -217,13 +217,38 @@ max(x::Real, y::Real) = max(promote(x,y)...)
...
-#   promote_op{R<:MyType,S<:MyType}(::typeof(*), ::Type{R}, ::Type{S}) = MyType{multype(R,S)}
-promote_op(::Any)    = (@_pure_meta; Bottom)
-promote_op(::Any, ::Any, ::Any...) = (@_pure_meta; Any)
-promote_op{T}(::Type{T}, ::Any) = (@_pure_meta; T)
+# "Promotion" that takes a function into account. These are meant to be
+# used mainly by broadcast methods, so it is adviced against overriding them
+if isdefined(Core, :Inference)
+    function _promote_op(op, T::Type)
+        G = Tuple{Generator{Tuple{T},typeof(op)}}
+        R = Core.Inference.return_type(first, G)
```
Comment (JeffBezanson) [inline base/promotion.jl]: Does `return_type(op, Tuple{T})` not work? This seems roundabout.

Comment (pabloferz) [inline base/promotion.jl]: That was the first thing I tried, the problem is that Test.@inferred does not work that way, but it works like this.
